scormxblock/scorm_file_uploader.py

```diff
@@ -0,0 +1,188 @@
+import os
+import tempfile
+import re
+import zipfile
+import shutil
+import logging
+
+from django.conf import settings
+from django.core.files.storage import default_storage
+from django.core.files.storage import get_storage_class
+from django.utils import encoding
+from django.core.cache import cache
+
+
+logger = logging.getLogger(__name__)
+
+# Regex to capture Content-Range header ranges.
+# e.g; "bytes 0-19999999/69221159"
+CONTENT_RE = re.compile(r"(?P<start>\d{1,11})-(?P<end>\d{1,11})/(?P<size>\d{1,11})")
+
+
+class FileAccessMode(object):
+    WRITE = "wb+"
+    APPEND = "ab+"
+
+
+class STATE(object):
+    """
+    enum for upload state
+    """
+    PROGRESS = 'progress'
+    ERROR = 'error'
+    COMPLETE = 'complete'
+
+
+class ScormPackageUploader(object):
+    """
+    Handles scorm package uploading
+    """
+    def __init__(self, request, xblock, scorm_storage_location):
+        self.xblock = xblock
+        self.request = request
+        self.scorm_file = request.params['scorm_file'].file
+        self.temp_file_path = os.path.join(tempfile.gettempdir(), xblock.location.block_id)
+        self.scorm_storage_location = os.path.join(scorm_storage_location, xblock.location.block_id)
+
+    def upload(self):
```

Comment:
I would again suggest to breakdown this function based on different upload status if possible. Try to analyse this function from test driven approach.